File: java/CountingSort.java
import java.util.Arrays;
import java.util.Collections;

class CountingSort {
	public static int max(int[] array) {
		int greatest = array[0];
		for (int i = 0; i < array.length; i++) {
			if (array[i] > greatest) {
				greatest = array[i];
			}
		}
		return greatest;
	}
	public static int[] sort(int[] input) {
		int greatest = max(input);
		int[] counts = new int[greatest + 1];
		for (int i = 0; i < input.length; i++) {
			counts[input[i]]++;
		}
		int counter = 0;
		for (int i = 0; i < counts.length; i++) {
			for (int j = 0; j < counts[i]; j++) {
				input[counter++] = i;
			}
		}
		return input;
	}
	public static void main(String[] args) {
		int[] input = new int[args.length];
		for(int i = 0; i < args.length; i++) {
			input[i] = Integer.parseInt(args[i]);
		}
		int[] output = sort(input);
		for(int i = 0; i < args.length; i++) {
			System.out.print("" + output[i] + " ");
		}
		System.out.println("");
	}
}

File: c/countingsort.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

int max(int* arr, size_t len) {
	if (len < 1) {
		return 0;
	}
	int greatest = arr[0];
	for (size_t i = 0; i < len; i++)
		if (arr[i] > greatest)
			greatest = arr[i];
	return greatest;
}

void sort(int* inputs, int* outputs, size_t length) {
	// this is the size of the array of counts
	int greatest = max(inputs, length);	// find the greatest number in the array

	// allocate the array of counts
	int* counts = calloc(greatest + 1, sizeof(int));
	if (counts == NULL) {
		fprintf(stderr, "Memory allocation error\n");
		exit(EXIT_FAILURE);
	}

	// count numbers in input array
	for (size_t i = 0; i < length; i++) {
		counts[inputs[i]]++;
	}

	size_t counter = 0;	// keep track of where we are in output array

	// loop through all the counts
	for (int i = 0; i <= greatest; i++) {	// for every count in array
		for (int j = 0; j < counts[i]; j++) {	// loop that many times
			outputs[counter++] = i;	// add the integer being counted to the output array
		}
	}
	free(counts);
}

int main(int argc, char** argv) {
	int *inputs, *outputs;
	size_t length = argc - 1;	// number of integers to sort

	if (argc < 2) {
		fprintf(stderr, "Not enough arguments given.\n");
		exit(EXIT_FAILURE);
	}

	inputs = calloc(length, sizeof(int));
	outputs = calloc(length, sizeof(int));

	if (inputs == NULL || outputs == NULL) {
		fprintf(stderr, "Memory allocation error\n");
		exit(EXIT_FAILURE);
	}

	for (int i = 1; i < argc; i++) {
		inputs[i - 1] = atoi(argv[i]);	// assign arguments to array
	}

	sort(inputs, outputs, length);

	for (size_t i = 0; i < length; i++) {
		printf("%d ", outputs[i]);	// print space separated sorted numbers
	}
	printf("\n");

	free(inputs);
	free(outputs);

	return EXIT_SUCCESS;
}
